tests/test_proxy_use.py

```diff
@@ -0,0 +1,366 @@
+    cls.http_proxy_addr = 'http://127.0.0.1:' + str(cls.http_proxy_port)
+
+
+    # Launch an HTTPS proxy server, also derived from inaz2/proxy2.
+    # (An HTTPS proxy performs its own TLS connection with the client and must
+    # be trusted by it, and is capable of tampering.)
+    # We instruct the proxy server to expect certain certificates from the
+    # target server.
+    # 1st arg: port
+    # 2nd arg: whether to intercept (HTTPS proxy) or relay (TCP tunnel using
+    #   HTTP CONNECT verb, to facilitate an HTTPS connection between the client
+    #   and server which the proxy cannot inspect)
+    # 3rd arg: (optional) certificate file for telling the proxy what target
+    #   server certs to accept in its HTTPS connection to the target server.
+    #   This is only relevant if the proxy is in intercept mode.
+    cls.https_proxy_port = cls.http_port + 3
+    cls.https_proxy_proc = subprocess.Popen(
+        ['python', 'proxy_server.py', str(cls.https_proxy_port), 'intercept',
+        os.path.join('ssl_certs', 'ssl_cert.crt')],
+        stderr=subprocess.PIPE)
+    # Note that the https proxy server's address uses https://, regardless of
+    # the type of connection used with the target server.
+    cls.https_proxy_addr = 'https://127.0.0.1:' + str(cls.https_proxy_port)
+
+    # Give the HTTP server and proxy server processes a little bit of time to
+    # start listening before allowing tests to begin, lest we get "Connection
+    # refused" errors. On the first test system. 0.1s was too short and 0.15s
+    # was long enough. Use 0.5s to be safe, and if issues arise, increase it.
+    time.sleep(0.5)
+
+
+
+
+
+  @classmethod
+  def tearDownClass(cls):
+    """
+    Cleanup performed after the last of the tests (TestWithProxies methods)
+    has been run.
+    Stop server process and perform clean up.
+    """
+    unittest_toolbox.Modified_TestCase.tearDownClass()
+
+    for proc in [
+        cls.http_server_proc,
+        cls.https_server_proc,
+        cls.http_proxy_proc,
+        cls.https_proxy_proc,
+      ]:
+      if proc.returncode is None:
+        logger.info('\tTerminating process ' + str(proc.pid) + ' in cleanup.')
+        proc.kill()
+
+
+
+  def setUp(self):
+    """
+    Setup performed before EACH test function (TestWithProxies class method)
+    runs.
+    """
+    unittest_toolbox.Modified_TestCase.setUp(self)
+
+    # Dictionary for saving environment values to restore.
+    self.old_env_values = {}
+
+    # Make a temporary file to serve on the server, and determine its length,
+    # and its url on the server.
+    current_dir = os.getcwd()
+    target_filepath = self.make_temp_data_file(directory=current_dir)
+    rel_target_filepath = os.path.basename(target_filepath)
+
+    with open(target_filepath, 'r') as target_file_object:
+      self.target_data_length = len(target_file_object.read())
+
+    self.url = \
+        'http://localhost:' + str(self.http_port) + '/' + rel_target_filepath
```

Comment:
Now `localhost`? ;-)

Reply:
The SSL certificates specify localhost (including one that's inconvenient to replace). It might be better to replace the 127.0.0.1 in the proxy server addresses with locahost, though.